src/cpp/src/speculative_decoding/speculative_decoding_stateful_eagle3.cpp

```diff
@@ -0,0 +1,1242 @@
+    OPENVINO_ASSERT(sample_count > 0 && sample_count <= shape[1],
+                    "Invalid sample_count: ",
+                    sample_count,
+                    ", logits seq_len: ",
+                    shape[1]);
+    OPENVINO_ASSERT(input_token_count > 0, "Invalid input_token_count");
+
+    const bool is_validation_mode = num_tokens_to_validate > 0;
+
+    log_debug(LogCategory::SAMPLE,
+              "sample_tokens: input_tokens=" + std::to_string(input_token_count) +
+                  ", sample_count=" + std::to_string(sample_count) +
+                  ", validate=" + std::to_string(num_tokens_to_validate) + ", logits_shape=" + format_shape(shape));
+
+    auto sequence_group = get_sequence_group();
+    OPENVINO_ASSERT(sequence_group, "SequenceGroup not initialized");
+
+    // TODO(top-k): Extend to support multiple sequences for top-k sampling
+    OPENVINO_ASSERT(get_running_sequence_count() == 1,
+                    "Eagle3 currently only supports top-1 sampling, got ",
+                    get_running_sequence_count(),
+                    " sequences");
+
+    auto current_seq = get_current_sequence();
+    OPENVINO_ASSERT(current_seq, "No running sequence at index 0");
+
+    const size_t prev_generated_len = current_seq->get_generated_len();
+    const size_t logits_seq_len = shape[1];
+    const size_t vocab_size = shape[2];
+
+    // Slice logits to last 'sample_count' positions if needed
+    ov::Tensor sliced_logits = logits;
+    if (sample_count < logits_seq_len) {
+        sliced_logits = ov::Tensor(logits, {0, logits_seq_len - sample_count, 0}, {1, logits_seq_len, vocab_size});
+    }
+
+    // Configure sequence group for sampling
+    sequence_group->schedule_tokens(input_token_count);
+    sequence_group->set_output_seq_len(sample_count);
+    sequence_group->set_num_validated_tokens(num_tokens_to_validate);
+
+    // Execute sampling
+    m_sampler.sample({sequence_group}, sliced_logits, is_validation_mode);
+    sequence_group->finish_iteration();
+
+    // Extract results based on mode
+    // TODO(top-k): For top-k, need to handle multiple sequences and their generated tokens
+    const auto& generated_ids = current_seq->get_generated_ids();
+    const size_t new_generated_len = generated_ids.size();
+
+    if (!is_validation_mode) {
+        OPENVINO_ASSERT(new_generated_len - prev_generated_len == sample_count,
+                        "Sampled token count mismatch: expected ",
+                        sample_count,
+                        ", got ",
+                        new_generated_len - prev_generated_len);
+
+        std::vector<int64_t> result_tokens(generated_ids.end() - sample_count, generated_ids.end());
+
+        log_debug(LogCategory::SAMPLE,
+                  "Sampled " + std::to_string(sample_count) + " token(s): " + format_tokens(result_tokens));
+
+        return (sample_count == 1) ? std::variant<int64_t, std::vector<int64_t>>(result_tokens[0]) : result_tokens;
```

Comment:
Do we really need explicit creation of `std::variant` here? In the similar code I used just return of either number of vector: https://github.com/openvinotoolkit/openvino.genai/blob/master/src/cpp/src/speculative_decoding/speculative_decoding_stateful.cpp#L345-L355